toc/rfc/rfc-draft-add-cnbs.md

```diff
@@ -0,0 +1,194 @@
+# Meta
+[meta]: #meta
+- Name: Add Cloud Native Buildpacks
+- Start Date: 2023-05-01
+- Author: @gerg
+- Contributing Authors: @ryanmoran, @robdimsdale, @dsabeti, @aramprice, @dsboulder
+- Status: Draft
+- RFC Pull Request: #591
+
+## Summary
+
+[Cloud Native Buildpacks (CNBs)](https://buildpacks.io/), also known as v3
+buildpacks, are the current generation of buildpacks, and offer some
+improvements over the v2 buildpacks that CF Deployment currently uses. The
+Cloud Foundry Foundation already has an implementation of Cloud Native
+Buildpacks via the [Paketo](https://paketo.io/) project, however these CNBs are
+not currently integrated with core CF Deployment.
+
+This RFC proposes adding CNB support to Cloud Foundry and including the Paketo
+buildpacks in CF Deployment.
+
+## Problem
+
+The v2 buildpacks are effectively in maintenance mode and do not receive
+substantial new features. By not integrating with v3 buildpacks, Cloud Foundry
+is missing out on new capabilities like application
+[SBOM](https://en.wikipedia.org/wiki/Software_supply_chain) generation, new
+buildpacks like the Java Native and Web Servers CNBs, as well as any new
+features that are added to the still-actively-developed v3 buildpacks.
+
+## Proposal
+
+### Overview
+
+Cloud Foundry integration with CNBs will be executed over at least two phases.
+
+In the first phase, the Paketo buildpacks will be "shimmed" to work with Cloud
+Foundry's existing v2 buildpack interface. These shimmed Paketo buildpacks will
+be bundled with select [CNB lifecycle](https://github.com/buildpacks/lifecycle)
+binaries and lightweight orchestration binaries. At build time, the Cloud
+Foundry [buildpack lifecycle](https://github.com/cloudfoundry/buildpackapplifecycle)
+will invoke the orchestration binaries, which will map the v2 build process to
+the v3 build process.
+
+In the second phase, unmodified Paketo buildpacks will be natively supported by
+Cloud Foundry. This will be implemented by integrating the buildpack shim logic
+into a Cloud Foundry lifecycle.
+
+Depending on learnings from the first two phases of CNB integration or other
+factors, there may be additional phases of work, however these additional
+phases are not covered by this RFC.
+
+Notably, this proposal does NOT include building or running OCI images on Cloud
+Foundry. This proposal is to use Paketo buildpacks to build and run Cloud
+Foundry droplets. This proposal does not preclude future efforts to integrate
+Cloud Foundry with OCI images, container registries, etc.
```

Comment:
I like incremental approaches in principle. One problem I see with this specific one is that we are heading ~180 degree in the other direction - if we wanted to follow up with pushing towards OCI images and container registries.
At least I don't see us reusing anything of the two phases laid out.

Reply:
From the system architecture perspective, this proposal is definitely divergent from building/running OCI images.  However, I think this proposal is convergent from the app developer perspective. My hypothesis is that app developers don't care that much about how their apps are packaged/run. With this proposal, app developers can get the benefits of the Paketo buildpacks, without requiring significant changes to the CF Deployment platform engineer/operator experience or system architecture.

Reply:
I agree with the app dev's point of view, but I also do see the platform engineer/operator view. I don't think that we should increase the system architecture's complexity with a shim in the mid- to long-term.

FWIW we need a CNB lifecycle run in a Diego scheduled container anyway, right? So the main addition to the system architecture would be a registry, right? I would assume that when [configured to use a blobstore like S3](https://github.com/distribution/distribution/blob/main/docs/configuration.md) this [registry implementation](https://github.com/distribution/distribution) will be mostly if not fully stateless.

Reply:
I think there is more needed than 'just a registry' to switch from droplets to OCI images. Droplets don't contain the stack and app developers are used to get stack updates automatically for buildpack apps without having to redeploy/restage their apps.
If the CNB lifecycle would produce OCI images, a stack rebase needs to be added to the runtime evacuation process if keep this user experience.